### ISSUE TITLE
Support for multiple APM servers in serverUrls config

### ISSUE THREAD

**Issue by gregkalapos** [elastic/apm-agent-dotnet#144]
From https://github.com/elastic/apm-agent-dotnet/pull/136#discussion_r264608577 

Maybe relevant: https://github.com/elastic/apm-agent-nodejs/issues/447

Java agent [server-urls](https://www.elastic.co/guide/en/apm/agent/java/current/config-reporter.html#config-server-urls) configuration setting supports multiple URLs which are used to implement fail-over behavior as follows
1. Randomly shuffle the list at the start. Shuffled list is the one to be used in the rest of the steps.
2. Set the first URL in the list as the current one.
3. Send data using the current URL.
4. If it fails set the next URL in the list as the current one and go to step 3.

See https://github.com/elastic/apm-agent-java/blob/master/apm-agent-core/src/main/java/co/elastic/apm/agent/report/IntakeV2ReportingEventHandler.java#L356 for details.

**Comment by SergeyKleyman** [comment]
Another approach is to follow node.js approach and rename the configuration setting from plural as it’s at the moment (`ServerUrls` and `ELASTIC_APM_SERVER_URLS`) to singular so it’s clear that we support only one server URL.

**Comment by SergeyKleyman** [comment]
@gregkalapos I've looked at the other Elastic APM agents and Java agent is the only one that support multiple values in `server-URL` configuration setting:
- [Java](https://www.elastic.co/guide/en/apm/agent/java/current/config-reporter.html#config-server-urls)
- [node.js](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#server-url)
- [Python](https://www.elastic.co/guide/en/apm/agent/python/current/configuration.html#config-server-url)
- [Ruby](https://www.elastic.co/guide/en/apm/agent/ruby/current/configuration.html#config-server-url)
- [Go](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-server-url)

Do you see a lot of value in this feature?
The easiest solution would be to rename it to `ELASTIC_APM_SERVER_URL` in 1.0 GA and bite a bullet on a breaking change.

**Comment by gregkalapos** [comment]
> Do you see a lot of value in this feature?

Definitely not a killer feature, it’s just a nice to have.

My only problem is that this is a major breaking change. If someone does not adapt the setting, the agent won’t send data to the server they configures with `ServerUrls`- In my opinion this is the worse type of breaking change. Of course we’d document this, but still I think it’s better to avoid these types of things.

I did not know that Java is the only agent that supports this, that’s definitely an interesting aspect.

But I still think it’s relatively low effort to implement the feature that the Java agent has and I’d stick to the original plan: keep it as it is in 1.0 GA - and mention this in the docs, and implement it for 1.1.

The downside of renaming it now is that once we implement it (and probably long term we’d implement it) we have to rename it again. That’d be 2 breaking changes - both could cause loss of data and lots of confusion just for fixing a small glitch in the name of this setting, which no user complained about until this point. (Or we keep both settings, which would be even more complicated) 

I agree that this is not ideal, but from a user’s point of view I think the alternative is worse and more hard to follow.

**Comment by gregkalapos** [comment]
The [Elastic APM Backend Agent Config Comparison](https://docs.google.com/spreadsheets/d/1JJjZotapacA3FkHc2sv_0wiChILi3uKnkwLTjtBmxwU/edit#gid=0) spreadsheet also says that `SERVER_URLS` is the official name of this config and none of the agents except Java and .NET use the official name currently - so probably the long term is to use plural - will check this assumption with the team.

**Comment by SergeyKleyman** [comment]
I've created #376 to handle the ServerUrls configuration option limitation (only one URL) for 1.0

**Comment by SergeyKleyman** [comment]
In addition to implementation the documentation will need to be updated:
- .NET agent setup instructions at https://www.elastic.co/products/apm
- [ServerUrls configuration option](https://www.elastic.co/guide/en/apm/agent/dotnet/current/config-reporter.html#config-server-urls)

**Comment by russcam** [comment]
Closing this as a consensus has been reached in https://github.com/elastic/apm/issues/359 to support only a single server url.
